git sync

akosyakov opened the pull request.

akosyakov commented:
@svenefftinge please review

kittaakos commented on packages/git/src/browser/git-sync-service.ts:
```diff
@@ -0,0 +1,227 @@
+    protected shouldPull(status: WorkingDirectoryStatus): boolean {
+        return status.aheadBehind && status.aheadBehind.behind > 0 || true;
+    }
+    protected async pull(repository: Repository, rebase: boolean): Promise<void> {
```
Wouldn't it make sense to move this and `push` to `git`?

akosyakov replied:
sounds good. 

I also had to reimplement `push` below, since duigite `push` api does not allow to provide `setUpsteam` and `force` flags.  Do you think it should be moved to the backend as well?

kittaakos replied:
Whatever changes you need in dugite-extra, let me know, and I make those. We can do it in the context of a follow-up, whichever is more convenient for you.

kittaakos commented on packages/git/src/browser/git-view-contribution.ts:
```diff
@@ -277,4 +285,42 @@ export class GitViewContribution extends AbstractViewContribution<GitWidget> {
+        const { upstreamBranch, aheadBehind } = status;
+        if (upstreamBranch) {
+            return {
+                text: '$(refresh)' + (aheadBehind ? ` ${aheadBehind.behind}↓ ${aheadBehind.ahead}↑` : ''),
```
That is a behavior change, isn't it? Previously we did not show ahead behind if both were 0. I am OK with this, just asking if the change was intentional. The `upstreamBranch` condition is also new.

akosyakov replied:
I've mimicked VS Code here. Do you think we should show?

kittaakos replied:
Let's stick to VSCode 👍. I just wanted to make sure that the change was made intentionally.

kittaakos replied:
Your changes make sense.

kittaakos commented on packages/git/src/browser/git-sync-service.ts:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    protected _syncing = false;
+    get syncing(): boolean {
```
Please switch to a method from the property getter if possible.

kittaakos replied:
I am not sure about this anymore... You know, it is fine that one could extend it, be then we do not use basic language features. It would be strange to have `setSync` `isSync`; you decide.

kittaakos commented on packages/git/src/browser/git-sync-service.ts:
```diff
@@ -0,0 +1,227 @@
+        return status.aheadBehind && status.aheadBehind.ahead > 0 || true;
+    }
+    protected shouldPull(status: WorkingDirectoryStatus): boolean {
+        return status.aheadBehind && status.aheadBehind.behind > 0 || true;
```
This is unused, and I think it is not correct. So when in detached mode, you do not have the `aheadBehind`, hence the expression will be `true`, but you cannot you should not pull because you are in detached mode.

akosyakov replied:
i will remove it.

svenefftinge commented:
When the local and remote branch are in sync it proposes with all three options.

<img width="641" alt="screen shot 2018-05-28 at 13 29 39" src="https://user-images.githubusercontent.com/372735/40612541-4e1648d6-627b-11e8-93ec-87e426f17b9a.png">

It should do nothing, no?

akosyakov commented:
@svenefftinge I've added the guard for this case.

@kittaakos i've added `rebase`, `setUpstream`, `force` flags to `pull` and `push` commands on `Git` interface. On the backend since dugite cannot handle them it fall backs to `exec` of one of these flags is enabled.

svenefftinge commented:
The spinning refresh icon isn'T replaced after I cancel an operation.

akosyakov commented:
@svenefftinge I've pushed a fix for https://github.com/theia-ide/theia/pull/1940#issuecomment-392724777. It spins now only when git operations are executed.
